packages/venia-concept/src/ServiceWorker/Utilities/__tests__/messageHandler.spec.js

```diff
@@ -0,0 +1,323 @@
+import {
+    __handlers__,
+    registerMessageHandler,
+    unRegisterMessageHandler,
+    handleMessageFromClient,
+    sendMessageToClient,
+    sendMessageToWindow
+} from '../messageHandler';
+
+const messageType = 'HELLO_WORLD';
+
+const handler = () => {};
+
+function clearHandlersObject() {
+    Object.keys(__handlers__).forEach(messageType => {
+        delete __handlers__[messageType];
+    });
+}
+
+afterEach(() => {
+    clearHandlersObject();
+});
+
+describe('Testing registerMessageHandler', () => {
+    test('registerMessageHandler should return function', () => {
+        expect(registerMessageHandler(messageType, handler)).toBeInstanceOf(
+            Function
+        );
+    });
+
+    test('registerMessageHandler should add a message handler for a given type in the handlers object', () => {
+        registerMessageHandler(messageType, handler);
+
+        expect(__handlers__).toHaveProperty(messageType);
+        expect(__handlers__[messageType]).toBeInstanceOf(Array);
+        expect(__handlers__[messageType]).toHaveLength(1);
+        expect(__handlers__[messageType]).toContain(handler);
+    });
```

Comment:
This may be too granular. The unit tests should test the input / output of the module and its functions. If someone comes along tomorrow and optimizes how `__handlers__` works (or any internals of the module), that shouldn't cause unit tests to fail.

I've found that approach to be the most scalable.

Reply:
See if there's any other way you can test that the rest of these test cases succeed without having to reach into `__handlers__` directly.

Reply:
I don't think that is possible. As far as I have researched online, the only way to test such `internal only/private` stuff it to either expose it with `__` so others know the risk or use [rewire](https://www.npmjs.com/package/rewire). I am not sure if we want to add a new npm dependency to test this. That is the reason why I settled with the first option.

Also, I feel, tests need to be granular enough to fail when their shape changes, in this case, they need to fail if the handlers are not stored as an `Array` because other functions expect it to be an `Array`.

Reply:
Yeah, I've used `rewire` before ... please, no 😅 

> Also, I feel, tests need to be granular enough to fail when their shape changes, in this case, they need to fail if the handlers are not stored as an Array because other functions expect it to be an Array.

If those other functions are internal to the module we don't have to explicitly care 😃 . The module will fail some test if has internal errors / inconsistencies.

My larger point is just that tests should rarely care about _how_ a module does its job, just that it does it. 

Tests for this module should be along the lines of "I registered multiple handlers, do they all get called when the event happens?". I think you already have these tests. If they work, we can assume the internal logic related to `__handlers__` is working properly.

Does that make sense?

Reply:
Or I could do something like this:

Assume there is a test mode like `process.env.TEST_MODE` set to true. In which case we can expose `__handlers__` by setting it in the `global` scope. This way we won't be leaking `__handlers__` in non-test environments.

What do you think about this option?

Reply:
> Does that make sense?

Yeah, it does. Will keep that in mind for future tests. Thanks.